File: packages/typediagram/src/tdbin/reader.ts
import { ok, type Result } from "../result.js";
import { tdbinErr } from "./error.js";
import {
  ELEM_BIT,
  ELEM_BYTE,
  ELEM_COMPOSITE,
  ELEM_EIGHT_BYTES,
  ELEM_POINTER,
  decodePointer,
  targetWord,
} from "./pointer.js";
import type { Pointer, Reader, StructCodec, TdbinError } from "./types.js";
import { verifyMessage } from "./verify.js";
import { WORD_BITS, WORD_BYTES, readWord, requireWordRange, utf8Decode } from "./word.js";

const MAX_DEPTH = 64;

interface CompositeList {
  readonly first: number;
  readonly count: number;
  readonly dataWords: number;
  readonly ptrWords: number;
  readonly stride: number;
}

export const message = <T>(codec: StructCodec<T>, bytes: Uint8Array): Result<T, TdbinError> => {
  if (bytes.length === 0 || bytes.length % WORD_BYTES !== 0) {
    return tdbinErr<T>("BadLength");
  }
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const verified = verifyMessage(bytes, view);
  if (!verified.ok) {
    return verified;
  }
  const head = readWord(bytes, view, 0);
  const ptr = head.ok ? decodePointer(head.value) : head;
  return ptr.ok ? readRoot(codec, bytes, view, ptr.value) : ptr;
};

export const scalar = (reader: Reader, at: number, slot: number): Result<bigint, TdbinError> =>
  slot >= reader.dataWords ? ok(0n) : readWord(reader.bytes, reader.view, at + slot);

export const boolBit = (reader: Reader, at: number, slot: number, bit: number): Result<boolean, TdbinError> => {
  const word = scalar(reader, at, slot);
  return word.ok && bit >= 0 && bit < WORD_BITS
    ? ok((word.value & (1n << BigInt(bit))) !== 0n)
    : tdbinErr<boolean>("LimitExceeded");
};

export const string = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<string | null, TdbinError> => {
  const raw = bytes(reader, at, _dataWords, slot);
  if (!raw.ok) {
    return raw;
  }
  if (raw.value === null) {
    return ok(null);
  }
  return utf8Decode(raw.value);
};

export const bytes = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<Uint8Array | null, TdbinError> => readBytes(reader, at, slot);

export const byteList = bytes;

export const boolList = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<boolean[] | null, TdbinError> => readList(reader, at, slot, ELEM_BIT, readBoolBody);

export const wordList = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<bigint[] | null, TdbinError> => readList(reader, at, slot, ELEM_EIGHT_BYTES, readWordBody);

export const bytes16List = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<readonly (readonly [bigint, bigint])[] | null, TdbinError> =>
  readComposite(reader, at, slot, readBytes16Body);

export const stringList = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<string[] | null, TdbinError> => readPointerList(reader, at, slot, readStringPointer);

export const bytesList = (
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number
): Result<Uint8Array[] | null, TdbinError> => readPointerList(reader, at, slot, readBytesPointer);

export const child = <T>(
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number,
  codec: StructCodec<T>
): Result<T | null, TdbinError> => {
  if (slot >= reader.ptrWords) {
    return ok(null);
  }
  const ptrWord = ptrIndex(reader, at, slot);
  const word = readWord(reader.bytes, reader.view, ptrWord);
  const ptr = word.ok ? decodePointer(word.value) : word;
  return ptr.ok ? readChildPointer(reader, ptrWord, codec, ptr.value) : ptr;
};

export const childList = <T>(
  reader: Reader,
  at: number,
  _dataWords: number,
  slot: number,
  codec: StructCodec<T>
): Result<T[] | null, TdbinError> => readComposite(reader, at, slot, (r, info) => readChildBody(r, info, codec));

export const requireNullPointer = (reader: Reader, at: number, slot: number): Result<void, TdbinError> => {
  if (slot >= reader.ptrWords) {
    return ok(undefined);
  }
  const ptrWord = ptrIndex(reader, at, slot);
  const word = readWord(reader.bytes, reader.view, ptrWord);
  const pointer = word.ok ? decodePointer(word.value) : word;
  return pointer.ok && pointer.value.kind === "null" ? ok(undefined) : tdbinErr("PointerKindMismatch");
};

const readRoot = <T>(
  codec: StructCodec<T>,
  source: Uint8Array,
  view: DataView,
  ptr: Pointer
): Result<T, TdbinError> => {
  if (ptr.kind !== "struct") {
    return ptr.kind === "null" ? tdbinErr<T>("NullRoot") : tdbinErr<T>("PointerKindMismatch");
  }
  const at = targetWord(0, ptr.offset);
  if (!at.ok) {
    return at;
  }
  const bounds = requireStructBounds(source, at.value, ptr.dataWords, ptr.ptrWords);
  if (!bounds.ok) {
    return bounds;
  }
  const reader = createReader(source, view, ptr.dataWords, ptr.ptrWords, source.length / WORD_BYTES);
  return codec.read(reader, at.value);
};

const createReader = (
  source: Uint8Array,
  view: DataView,
  dataWords: number,
  ptrWords: number,
  wordCount: number
): Reader => ({
  bytes: source,
  view,
  dataWords,
  ptrWords,
  depth: MAX_DEPTH,
  budget: { value: wordCount },
});

const ptrIndex = (reader: Reader, at: number, slot: number): number => at + reader.dataWords + slot;

const readBytes = (reader: Reader, at: number, slot: number): Result<Uint8Array | null, TdbinError> => {
  if (slot >= reader.ptrWords) {
    return ok(null);
  }
  const ptrWord = ptrIndex(reader, at, slot);
  const word = readWord(reader.bytes, reader.view, ptrWord);
  const ptr = word.ok ? decodePointer(word.value) : word;
  return ptr.ok ? readBytesPointerKind(reader, ptrWord, ptr.value) : ptr;
};

const readList = <T>(
  reader: Reader,
  at: number,
  slot: number,
  elem: number,
  readBody: (reader: Reader, ptrWord: number, offset: number, count: number) => Result<T, TdbinError>
): Result<T | null, TdbinError> => {
  if (slot >= reader.ptrWords) {
    return ok(null);
  }
  const ptrWord = ptrIndex(reader, at, slot);
  const word = readWord(reader.bytes, reader.view, ptrWord);
  const ptr = word.ok ? decodePointer(word.value) : word;
  return ptr.ok ? readListPointer(reader, ptrWord, elem, readBody, ptr.value) : ptr;
};

const readComposite = <T>(
  reader: Reader,
  at: number,
  slot: number,
  readBody: (reader: Reader, info: CompositeList) => Result<T, TdbinError>
): Result<T | null, TdbinError> =>
  readList<T>(reader, at, slot, ELEM_COMPOSITE, (r, ptrWord, offset, count) => {
    const header = readCompositeHeader(r, ptrWord, offset, count);
    return header.ok ? readBody(r, header.value) : header;
  });

const readPointerList = <T>(
  reader: Reader,
  at: number,
  slot: number,
  readOne: (reader: Reader, ptrWord: number) => Result<T, TdbinError>
): Result<T[] | null, TdbinError> =>
  readList<T[]>(reader, at, slot, ELEM_POINTER, (r, ptrWord, offset, count) =>
    readPointerBody(r, ptrWord, offset, count, readOne)
  );

const readChildPointer = <T>(
  reader: Reader,
  ptrWord: number,
  codec: StructCodec<T>,
  ptr: Pointer
): Result<T | null, TdbinError> => {
  if (ptr.kind === "null") {
    return ok(null);
  }
  return ptr.kind === "struct" ? followStruct(reader, ptrWord, codec, ptr) : tdbinErr<T | null>("PointerKindMismatch");
};

const readBytesPointerKind = (reader: Reader, ptrWord: number, ptr: Pointer): Result<Uint8Array | null, TdbinError> => {
  if (ptr.kind === "null") {
    return ok(null);
  }
  return ptr.kind === "list" && ptr.elem === ELEM_BYTE
    ? readListBytes(reader, ptrWord, ptr.offset, ptr.count)
    : tdbinErr<Uint8Array | null>("PointerKindMismatch");
};

const readListPointer = <T>(
  reader: Reader,
  ptrWord: number,
  elem: number,
  readBody: (reader: Reader, ptrWord: number, offset: number, count: number) => Result<T, TdbinError>,
  ptr: Pointer
): Result<T | null, TdbinError> => {
  if (ptr.kind === "null") {
    return ok(null);
  }
  return ptr.kind === "list" && ptr.elem === elem
    ? readBody(reader, ptrWord, ptr.offset, ptr.count)
    : tdbinErr<T | null>("PointerKindMismatch");
};

const followStruct = <T>(
  reader: Reader,
  ptrWord: number,
  codec: StructCodec<T>,
  ptr: Extract<Pointer, { kind: "struct" }>
): Result<T, TdbinError> => {
  const target = targetWord(ptrWord, ptr.offset);
  if (!target.ok) {
    return target;
  }
  const bounds = requireStructBounds(reader.bytes, target.value, ptr.dataWords, ptr.ptrWords);
  if (!bounds.ok) {
    return bounds;
  }
  const childReader = descend(reader, ptr.dataWords, ptr.ptrWords);
  return childReader.ok ? codec.read(childReader.value, target.value) : childReader;
};

const descend = (reader: Reader, dataWords: number, ptrWords: number): Result<Reader, TdbinError> => {
  if (reader.depth <= 0) {
    return tdbinErr("DepthExceeded");
  }
  if (reader.budget.value <= 0) {
    return tdbinErr("AmplificationExceeded");
  }
  reader.budget.value = reader.budget.value - 1;
  return ok({ ...reader, dataWords, ptrWords, depth: reader.depth - 1 });
};

const readListBytes = (
  reader: Reader,
  ptrWord: number,
  offset: number,
  count: number
): Result<Uint8Array, TdbinError> => {
  const startWord = targetWord(ptrWord, offset);
  if (!startWord.ok) {
    return startWord;
  }
  const start = startWord.value * WORD_BYTES;
  const end = start + count;
  return end <= reader.bytes.length
    ? ok(reader.bytes.slice(start, end))
    : tdbinErr("PointerOutOfBounds", { wordIndex: ptrWord });
};

const readBoolBody = (
  reader: Reader,
  ptrWord: number,
  offset: number,
  count: number
): Result<boolean[], TdbinError> => {
  const start = targetWord(ptrWord, offset);
  if (!start.ok) {
    return start;
  }
  const words = Math.ceil(count / WORD_BITS);
  const bounds = requireWordRange(reader.bytes, start.value, words);
  return bounds.ok ? unpackBools(reader, start.value, count) : bounds;
};

const readWordBody = (reader: Reader, ptrWord: number, offset: number, count: number): Result<bigint[], TdbinError> => {
  const start = targetWord(ptrWord, offset);
  if (!start.ok) {
    return start;
  }
  const bounds = requireWordRange(reader.bytes, start.value, count);
  return bounds.ok ? readWords(reader, start.value, count) : bounds;
};

const readPointerBody = <T>(
  reader: Reader,
  ptrWord: number,
  offset: number,
  count: number,
  readOne: (reader: Reader, ptrWord: number) => Result<T, TdbinError>
): Result<T[], TdbinError> => {
  const start = targetWord(ptrWord, offset);
  if (!start.ok) {
    return start;
  }
  const bounds = requireWordRange(reader.bytes, start.value, count);
  return bounds.ok ? readPointerItems(reader, start.value, count, readOne) : bounds;
};

const readStringPointer = (reader: Reader, ptrWord: number): Result<string, TdbinError> => {
  const raw = readBytesPointer(reader, ptrWord);
  return raw.ok ? utf8Decode(raw.value) : raw;
};

const readBytesPointer = (reader: Reader, ptrWord: number): Result<Uint8Array, TdbinError> => {
  const word = readWord(reader.bytes, reader.view, ptrWord);
  const ptr = word.ok ? decodePointer(word.value) : word;
  const raw = ptr.ok ? readBytesPointerKind(reader, ptrWord, ptr.value) : ptr;
  return raw.ok ? ok(raw.value ?? new Uint8Array()) : raw;
};

const readBytes16Body = (
  reader: Reader,
  info: CompositeList
): Result<readonly (readonly [bigint, bigint])[], TdbinError> => {
  if (info.dataWords !== 2 || info.ptrWords !== 0) {
    return tdbinErr<readonly (readonly [bigint, bigint])[]>("PointerKindMismatch");
  }
  return readBytes16Items(reader, info);
};

const readChildBody = <T>(reader: Reader, info: CompositeList, codec: StructCodec<T>): Result<T[], TdbinError> =>
  readChildItems(reader, info, codec);

const readChildItems = <T>(reader: Reader, info: CompositeList, codec: StructCodec<T>): Result<T[], TdbinError> => {
  const values: T[] = [];
  for (let index = 0; index < info.count; index += 1) {
    const childValue = readInlineStruct(reader, elemAt(info, index), info, codec);
    if (!childValue.ok) {
      return childValue;
    }
    values.push(childValue.value);
  }
  return ok(values);
};

const readCompositeHeader = (
  reader: Reader,
  ptrWord: number,
  offset: number,
  elemWords: number
): Result<CompositeList, TdbinError> => {
  const tagAt = targetWord(ptrWord, offset);
  const word = tagAt.ok ? readWord(reader.bytes, reader.view, tagAt.value) : tagAt;
  const ptr = word.ok ? decodePointer(word.value) : word;
  return tagAt.ok && ptr.ok && ptr.value.kind === "struct"
    ? compositeInfo(reader.bytes, tagAt.value, ptr.value.offset, ptr.value.dataWords, ptr.value.ptrWords, elemWords)
    : tdbinErr<CompositeList>("PointerKindMismatch");
};

const compositeInfo = (
  source: Uint8Array,
  tagAt: number,
  count: number,
  dataWords: number,
  ptrWords: number,
  elemWords: number
): Result<CompositeList, TdbinError> => {
  const stride = dataWords + ptrWords;
  const expected = stride * count;
  const valid = expected === elemWords && (stride !== 0 || count === 0);
  if (!valid) {
    return tdbinErr<CompositeList>("MalformedCompositeTag");
  }
  const bounds = requireWordRange(source, tagAt, expected + 1);
  return bounds.ok ? ok({ first: tagAt + 1, count, dataWords, ptrWords, stride }) : bounds;
};

const readInlineStruct = <T>(
  reader: Reader,
  at: number,
  info: CompositeList,
  codec: StructCodec<T>
): Result<T, TdbinError> => {
  const bounds = requireStructBounds(reader.bytes, at, info.dataWords, info.ptrWords);
  if (!bounds.ok) {
    return bounds;
  }
  const childReader = descend(reader, info.dataWords, info.ptrWords);
  return childReader.ok ? codec.read(childReader.value, at) : childReader;
};

const unpackBools = (reader: Reader, start: number, count: number): Result<boolean[], TdbinError> => {
  const values: boolean[] = [];
  for (let wordIndex = 0; wordIndex < Math.ceil(count / WORD_BITS); wordIndex += 1) {
    const word = readWord(reader.bytes, reader.view, start + wordIndex);
    if (!word.ok) {
      return word;
    }
    const remaining = Math.min(WORD_BITS, count - wordIndex * WORD_BITS);
    for (let bit = 0; bit < remaining; bit += 1) {
      values.push((word.value & (1n << BigInt(bit))) !== 0n);
    }
  }
  return ok(values);
};

const readWords = (reader: Reader, start: number, count: number): Result<bigint[], TdbinError> => {
  const values: bigint[] = [];
  for (let index = 0; index < count; index += 1) {
    const word = readWord(reader.bytes, reader.view, start + index);
    if (!word.ok) {
      return word;
    }
    values.push(word.value);
  }
  return ok(values);
};

const readPointerItems = <T>(
  reader: Reader,
  start: number,
  count: number,
  readOne: (reader: Reader, ptrWord: number) => Result<T, TdbinError>
): Result<T[], TdbinError> => {
  const values: T[] = [];
  for (let index = 0; index < count; index += 1) {
    const item = readOne(reader, start + index);
    if (!item.ok) {
      return item;
    }
    values.push(item.value);
  }
  return ok(values);
};

const readBytes16Items = (
  reader: Reader,
  info: CompositeList
): Result<readonly (readonly [bigint, bigint])[], TdbinError> => {
  const values: Array<readonly [bigint, bigint]> = [];
  for (let index = 0; index < info.count; index += 1) {
    const at = elemAt(info, index);
    const first = readWord(reader.bytes, reader.view, at);
    if (!first.ok) {
      return first;
    }
    const second = readWord(reader.bytes, reader.view, at + 1);
    if (!second.ok) {
      return second;
    }
    values.push([first.value, second.value]);
  }
  return ok(values);
};

const elemAt = (info: CompositeList, index: number): number => info.first + info.stride * index;

const requireStructBounds = (source: Uint8Array, at: number, dataWords: number, ptrWords: number) =>
  requireWordRange(source, at, dataWords + ptrWords);

File: crates/tdbin/src/reader.rs
//! The fused single-pass decoder/verifier ([TDBIN-SAFE]): every pointer the
//! typed reader follows is bounds-checked, depth-capped, and charged to the
//! amplification budget, and pointer slots the schema does not visit are
//! walked by the structural verifier ([TDBIN-REC-SHORT],
//! [TDBIN-UNION-UNKNOWN]) — so one traversal both validates and materializes.
//! List decoding lives in `reader_lists.rs`; columnar decoding in `column.rs`.

use core::cell::Cell;

use crate::error::DecodeError;
use crate::layout::{self, WORD_BYTES};
use crate::pointer::{self, Pointer, ELEM_BYTE};
use crate::{Struct, MAX_DEPTH};

/// A bounds-checked view over an encoded message ([TDBIN-SAFE-BOUNDS]).
#[derive(Debug)]
pub struct Reader<'a> {
    /// The complete message bytes.
    bytes: &'a [u8],
    /// Actual data-section words declared by the struct pointer.
    data_words: u16,
    /// Actual pointer-section words declared by the struct pointer.
    ptr_words: u16,
    /// Remaining struct-nesting depth ([TDBIN-SAFE-DEPTH]).
    depth: u32,
    /// Shared remaining traversal budget in words ([TDBIN-SAFE-AMPLIFY]).
    budget: &'a Cell<u64>,
    /// Shared remaining materialization budget in rows/values
    /// ([TDBIN-COL-SAFE]).
    materialize: &'a Cell<u64>,
}

/// Absolute per-message materialization budget: decoded rows and block values
/// may not exceed this, whatever the wire claims ([TDBIN-COL-SAFE]).
const MATERIALIZE_BUDGET: u64 = 1 << 26;

impl<'a> Reader<'a> {
    /// Decode a root value from a complete message ([TDBIN-MSG-BARE]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] on malformed or out-of-bounds input.
    pub(crate) fn message<T: Struct>(bytes: &[u8]) -> Result<T, DecodeError> {
        let len = bytes.len();
        ((len != 0) && len.is_multiple_of(WORD_BYTES))
            .then_some(())
            .ok_or(DecodeError::BadLength)?;
        let words = u64::try_from(len / WORD_BYTES).map_err(|_| DecodeError::BadLength)?;
        let budget = Cell::new(words.checked_sub(1).ok_or(DecodeError::BadLength)?);
        let materialize = Cell::new(MATERIALIZE_BUDGET);
        let head = layout::read_word(bytes, 0)?;
        match pointer::decode(head)? {
            Pointer::Struct {
                offset,
                data_words,
                ptr_words,
            } => Self::read_root(bytes, &budget, &materialize, offset, data_words, ptr_words),
            Pointer::Null => Err(DecodeError::NullRoot),
            Pointer::List { .. } => Err(DecodeError::PointerKindMismatch),
        }
    }

    /// Follow the root struct pointer and materialize the root value.
    fn read_root<T: Struct>(
        bytes: &[u8],
        budget: &Cell<u64>,
        materialize: &Cell<u64>,
        offset: i64,
        data_words: u16,
        ptr_words: u16,
    ) -> Result<T, DecodeError> {
        let at =
            layout::target(0, offset).ok_or(DecodeError::PointerOutOfBounds { word_index: 0 })?;
        Self::require_struct_bounds(bytes, at, data_words, ptr_words)?;
        let reader = Reader {
            bytes,
            data_words,
            ptr_words,
            depth: MAX_DEPTH,
            budget,
            materialize,
        };
        reader.charge(section_words(data_words, ptr_words)?)?;
        reader.read_struct_verified::<T>(at)
    }

    /// Read a scalar data word from `slot` of the struct at `at`.
    ///
    /// # Errors
    /// Returns [`DecodeError`] if the word is out of bounds.
    pub fn scalar(&self, at: usize, slot: u16) -> Result<u64, DecodeError> {
        if slot >= self.data_words {
            return Ok(0);
        }
        let idx = at
            .checked_add(usize::from(slot))
            .ok_or(DecodeError::PointerOutOfBounds { word_index: at })?;
        layout::read_word(self.bytes, idx)
    }

    /// Read one packed Bool bit from a data word ([TDBIN-REC-XOR]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] if the slot is out of bounds or the bit is invalid.
    pub fn bool_bit(&self, at: usize, slot: u16, bit: u8) -> Result<bool, DecodeError> {
        let mask = 1_u64
            .checked_shl(u32::from(bit))
            .ok_or(DecodeError::LimitExceeded)?;
        self.scalar(at, slot).map(|word| word & mask != 0)
    }

    /// Read an optional UTF-8 string from pointer `slot` ([TDBIN-PRIM-MAP]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] on malformed input or invalid UTF-8.
    pub fn string(
        &self,
        at: usize,
        _data_words: u16,
        slot: u16,
    ) -> Result<Option<String>, DecodeError> {
        match self.read_bytes(at, slot)? {
            None => Ok(None),
            Some(raw) => String::from_utf8(raw)
                .map(Some)
                .map_err(|_| DecodeError::InvalidUtf8),
        }
    }

    /// Read an optional raw byte list from pointer `slot` ([TDBIN-PRIM-MAP]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] on malformed input.
    pub fn bytes(
        &self,
        at: usize,
        _data_words: u16,
        slot: u16,
    ) -> Result<Option<Vec<u8>>, DecodeError> {
        self.read_bytes(at, slot)
    }

    /// Read an optional child struct from pointer `slot` ([TDBIN-PTR-STRUCT]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] on malformed input, depth, or amplification.
    pub fn child<C: Struct>(
        &self,
        at: usize,
        _data_words: u16,
        slot: u16,
    ) -> Result<Option<C>, DecodeError> {
        if slot >= self.ptr_words {
            return Ok(None);
        }
        let ptr_word = self.ptr_index(at, slot)?;
        match pointer::decode(layout::read_word(self.bytes, ptr_word)?)? {
            Pointer::Null => Ok(None),
            Pointer::Struct {
                offset,
                data_words,
                ptr_words,
            } => self
                .follow_struct::<C>(ptr_word, offset, data_words, ptr_words)
                .map(Some),
            Pointer::List { .. } => Err(DecodeError::PointerKindMismatch),
        }
    }

    /// Require an inactive union pointer slot to remain null.
    ///
    /// # Errors
    /// Returns [`DecodeError::PointerKindMismatch`] for a non-null slot.
    pub fn require_null_pointer(&self, at: usize, slot: u16) -> Result<(), DecodeError> {
        if slot >= self.ptr_words {
            return Ok(());
        }
        let ptr_word = self.ptr_index(at, slot)?;
        match pointer::decode(layout::read_word(self.bytes, ptr_word)?)? {
            Pointer::Null => Ok(()),
            Pointer::Struct { .. } | Pointer::List { .. } => Err(DecodeError::PointerKindMismatch),
        }
    }

    /// Structurally verify every actual pointer slot of the struct at `at`.
    ///
    /// Generated union decoders call this before surfacing `UnknownVariant`,
    /// so a message with an unknown discriminant is still fully
    /// pointer-checked ([TDBIN-UNION-UNKNOWN], [TDBIN-SAFE-ZEROSLOT]).
    ///
    /// # Errors
    /// Returns [`DecodeError`] on any structurally invalid slot.
    pub fn verify_struct_slots(&self, at: usize) -> Result<(), DecodeError> {
        self.verify_slots_from(at, 0)
    }

    /// Read a typed struct then verify its schema-unknown extension slots.
    pub(crate) fn read_struct_verified<C: Struct>(&self, at: usize) -> Result<C, DecodeError> {
        let value = C::read_struct(self, at)?;
        self.verify_slots_from(at, C::PTR_WORDS)?;
        Ok(value)
    }

    /// Structurally verify actual pointer slots starting at `from`.
    pub(crate) fn verify_slots_from(&self, at: usize, from: u16) -> Result<(), DecodeError> {
        for slot in from..self.ptr_words {
            let idx = self.ptr_index(at, slot)?;
            crate::verify::pointer_word(self.bytes, idx, self.depth, self.budget)?;
        }
        Ok(())
    }

    /// Read an optional raw byte list from pointer `slot`.
    pub(crate) fn read_bytes(&self, at: usize, slot: u16) -> Result<Option<Vec<u8>>, DecodeError> {
        if slot >= self.ptr_words {
            return Ok(None);
        }
        let ptr_word = self.ptr_index(at, slot)?;
        match pointer::decode(layout::read_word(self.bytes, ptr_word)?)? {
            Pointer::Null => Ok(None),
            Pointer::List {
                offset,
                elem,
                count,
            } => self
                .read_byte_slice(ptr_word, offset, elem, count)
                .map(|slice| Some(slice.to_vec())),
            Pointer::Struct { .. } => Err(DecodeError::PointerKindMismatch),
        }
    }

    /// Borrow the `count` bytes referenced by a byte-list pointer.
    pub(crate) fn read_byte_slice(
        &self,
        ptr_word: usize,
        offset: i64,
        elem: u8,
        count: u32,
    ) -> Result<&'a [u8], DecodeError> {
        if elem != ELEM_BYTE {
            return Err(DecodeError::PointerKindMismatch);
        }
        let start_word = Self::list_start(ptr_word, offset)?;
        let len = usize::try_from(count).map_err(|_| DecodeError::LimitExceeded)?;
        Self::require_word_range(self.bytes, start_word, len.div_ceil(WORD_BYTES))?;
        self.charge(len.div_ceil(WORD_BYTES))?;
        let start = start_word
            .checked_mul(WORD_BYTES)
            .ok_or(DecodeError::PointerOutOfBounds {
                word_index: start_word,
            })?;
        let end = start
            .checked_add(len)
            .ok_or(DecodeError::PointerOutOfBounds {
                word_index: start_word,
            })?;
        self.bytes
            .get(start..end)
            .ok_or(DecodeError::PointerOutOfBounds {
                word_index: start_word,
            })
    }

    /// Follow a struct pointer, enforcing depth and amplification budgets.
    pub(crate) fn follow_struct<C: Struct>(
        &self,
        ptr_word: usize,
        offset: i64,
        data_words: u16,
        ptr_words: u16,
    ) -> Result<C, DecodeError> {
        let target = layout::target(ptr_word, offset).ok_or(DecodeError::PointerOutOfBounds {
            word_index: ptr_word,
        })?;
        Self::require_struct_bounds(self.bytes, target, data_words, ptr_words)?;
        let child = self.descend(data_words, ptr_words)?;
        child.charge(section_words(data_words, ptr_words)?)?;
        child.read_struct_verified::<C>(target)
    }

    /// Read one inline composite element (its words are already charged).
    pub(crate) fn read_inline_struct<C: Struct>(
        &self,
        target: usize,
        data_words: u16,
        ptr_words: u16,
    ) -> Result<C, DecodeError> {
        Self::require_struct_bounds(self.bytes, target, data_words, ptr_words)?;
        let child = self.descend(data_words, ptr_words)?;
        child.read_struct_verified::<C>(target)
    }

    /// A child view one nesting level down with the given wire sections.
    pub(crate) fn descend(
        &self,
        data_words: u16,
        ptr_words: u16,
    ) -> Result<Reader<'a>, DecodeError> {
        let depth = self
            .depth
            .checked_sub(1)
            .ok_or(DecodeError::DepthExceeded)?;
        Ok(Reader {
            bytes: self.bytes,
            data_words,
            ptr_words,
            depth,
            budget: self.budget,
            materialize: self.materialize,
        })
    }

    /// Deduct `amount` from a budget cell, failing once it would go negative.
    /// Shared by the amplification and materialization budgets.
    fn charge_cell(cell: &Cell<u64>, amount: usize) -> Result<(), DecodeError> {
        let cost = u64::try_from(amount).map_err(|_| DecodeError::LimitExceeded)?;
        let left = cell
            .get()
            .checked_sub(cost)
            .ok_or(DecodeError::AmplificationExceeded)?;
        cell.set(left);
        Ok(())
    }

    /// Charge traversed words to the amplification budget ([TDBIN-SAFE-AMPLIFY]).
    pub(crate) fn charge(&self, words: usize) -> Result<(), DecodeError> {
        Self::charge_cell(self.budget, words)
    }

    /// Charge decoded rows or block values against the absolute
    /// materialization budget ([TDBIN-COL-SAFE]).
    pub(crate) fn charge_materialized(&self, rows: usize) -> Result<(), DecodeError> {
        Self::charge_cell(self.materialize, rows)
    }

    /// The message bytes this reader validates against.
    pub(crate) fn wire(&self) -> &'a [u8] {
        self.bytes
    }

    /// Actual pointer-section width declared by the followed struct pointer.
    pub(crate) fn wire_ptr_words(&self) -> u16 {
        self.ptr_words
    }

    /// Absolute word index of pointer `slot`.
    pub(crate) fn ptr_index(&self, at: usize, slot: u16) -> Result<usize, DecodeError> {
        layout::ptr_word(at, self.data_words, usize::from(slot))
            .ok_or(DecodeError::PointerOutOfBounds { word_index: at })
    }

    /// Resolve a list pointer target.
    pub(crate) fn list_start(ptr_word: usize, offset: i64) -> Result<usize, DecodeError> {
        layout::target(ptr_word, offset).ok_or(DecodeError::PointerOutOfBounds {
            word_index: ptr_word,
        })
    }

    /// Require a struct body to fit inside the message.
    pub(crate) fn require_struct_bounds(
        bytes: &[u8],
        at: usize,
        data_words: u16,
        ptr_words: u16,
    ) -> Result<(), DecodeError> {
        let words = section_words(data_words, ptr_words)?;
        Self::require_word_range(bytes, at, words)
    }

    /// Require `words` words starting at `at` to fit inside the message.
    pub(crate) fn require_word_range(
        bytes: &[u8],
        at: usize,
        words: usize,
    ) -> Result<(), DecodeError> {
        let end = at.checked_add(words).ok_or(DecodeError::LimitExceeded)?;
        (end <= bytes.len() / WORD_BYTES)
            .then_some(())
            .ok_or(DecodeError::PointerOutOfBounds { word_index: at })
    }
}

/// Return a struct's total section width.
fn section_words(data_words: u16, ptr_words: u16) -> Result<usize, DecodeError> {
    usize::from(data_words)
        .checked_add(usize::from(ptr_words))
        .ok_or(DecodeError::LimitExceeded)
}
